domain/upgrade/state/state.go

```diff
@@ -134,58 +141,107 @@ AND    upgrade_node.controller_node_id IS NULL;
 		if err != nil {
 			return errors.Trace(err)
 		}
+		defer rows.Close()
+
 		for rows.Next() {
 			var unreadyControllers int
-			err := rows.Scan(&unreadyControllers)
+			err = rows.Scan(&unreadyControllers)
 			if err != nil {
 				return errors.Trace(err)
 			}
 			allReady = unreadyControllers == 0
 		}
-		return nil
+
+		return rows.Err()
 	})
 	if err != nil {
 		return false, errors.Trace(err)
 	}
 	return allReady, nil
 }
 
-// StartUpgrade starts the provided upgrade is it exists
+// StartUpgrade starts the provided upgrade if the upgrade already exists.
+// If it's already started, it becomes a no-op.
 //
 // TODO (jack-w-shaw) Set `statuses`/`statuseshistory` here
 // to status.Busy once the table has been added
-func (st *State) StartUpgrade(ctx context.Context, upgradeUUID string) error {
+func (st *State) StartUpgrade(ctx context.Context, upgradeUUID domainupgrade.UUID) error {
 	db, err := st.DB()
 	if err != nil {
 		return errors.Trace(err)
 	}
 
-	getUpgradeStartedQuery := "SELECT started_at AS &info.* FROM upgrade_info WHERE uuid = $M.info_uuid"
-	getUpgradeStartedStmt, err := sqlair.Prepare(getUpgradeStartedQuery, info{}, sqlair.M{})
+	getUpgradeStartedQuery := `
+SELECT upgrade_state_type.id 
+FROM upgrade_info 
+	LEFT JOIN upgrade_state_type
+		ON upgrade_info.state_type_id = upgrade_state_type.id
+WHERE uuid = ?
+`
+
+	startUpgradeQuery := "UPDATE upgrade_info SET state_type_id = ? WHERE uuid = ? AND state_type_id = ?"
+
+	return errors.Trace(db.StdTxn(ctx, func(ctx context.Context, tx *sql.Tx) error {
+		var state int
+		row := tx.QueryRowContext(ctx, getUpgradeStartedQuery, upgradeUUID)
+		if err := row.Scan(&state); err != nil {
+			return errors.Trace(err)
+		}
+
+		// If the upgrade is already started, we don't need to do anything.
+		if err := upgrade.State(state).TransitionTo(upgrade.Started); err != nil {
```

Comment:
This is where the state-machine isn't so great. Migrations have a single actor, but here we have multiple actors where if someone starts an already-started upgrade, it's clearly an error condition; not a carry-on condition.

Reply:
One of the prior tests stated that a start upgrade called twice on the same upgrade was idempotent, which is why I returned nil here.

https://github.com/juju/juju/blob/5b9c1acd0e502fd54545332dd4198d44b2c4d899/domain/upgrade/state/state_test.go#L200C22-L218

Reply:
I think that test was wrong. You shouldn't be able to call start upgrade twice and return nothing. Returning a sentinel error for when this occurs, so it's possible to then call `UpgradeInfo` to find out what state everything is in...

We have to remember that if a step has failed, it should not be possible to undo that, without manual intervention; we don't know what the state of the world is at. If we can't progress this action, then there is no automatic way to undo this.